File: livetrace/chrome-console-monitor.js
#!/usr/bin/env node

const WebSocket = require('ws');
const http = require('http');
const fs = require('fs');

const LOG_FILE = '/tmp/chrome-live-console.log';
const DEBUG_PORT = 9222;

// Clear the log file
fs.writeFileSync(LOG_FILE, `Chrome Console Monitor Started - ${new Date().toISOString()}\n${'='.repeat(80)}\n\n`);

function log(message) {
  const timestamp = new Date().toISOString();
  const logLine = `[${timestamp}] ${message}\n`;
  fs.appendFileSync(LOG_FILE, logLine);
  console.log(logLine.trim());
}

async function getWebSocketDebuggerUrl() {
  return new Promise((resolve, reject) => {
    http.get(`http://localhost:${DEBUG_PORT}/json`, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const tabs = JSON.parse(data);
          const tab = tabs.find(t => t.type === 'page') || tabs[0];
          if (tab && tab.webSocketDebuggerUrl) {
            resolve(tab.webSocketDebuggerUrl);
          } else {
            reject(new Error('No debugger URL found'));
          }
        } catch (err) {
          reject(err);
        }
      });
    }).on('error', reject);
  });
}

async function main() {
  try {
    log('Connecting to Chrome DevTools Protocol...');

    const wsUrl = await getWebSocketDebuggerUrl();
    log(`WebSocket URL: ${wsUrl}`);

    const ws = new WebSocket(wsUrl);

    ws.on('open', () => {
      log('Connected to Chrome DevTools');

      // Enable Console domain
      ws.send(JSON.stringify({ id: 1, method: 'Console.enable' }));

      // Enable Runtime domain for console messages
      ws.send(JSON.stringify({ id: 2, method: 'Runtime.enable' }));

      // Enable Log domain
      ws.send(JSON.stringify({ id: 3, method: 'Log.enable' }));

      // Enable Network domain for network logs
      ws.send(JSON.stringify({ id: 4, method: 'Network.enable' }));

      log('Console monitoring enabled. Waiting for messages...\n');
    });

    ws.on('message', (data) => {
      try {
        const message = JSON.parse(data);

        // Console API messages (console.log, console.error, etc.)
        if (message.method === 'Runtime.consoleAPICalled') {
          const { type, args, stackTrace } = message.params;
          const logArgs = args.map(arg => {
            if (arg.value !== undefined) return arg.value;
            if (arg.description) return arg.description;
            return JSON.stringify(arg);
          }).join(' ');

          log(`[CONSOLE.${type.toUpperCase()}] ${logArgs}`);

          if (stackTrace) {
            log(`  Stack: ${stackTrace.callFrames[0]?.url}:${stackTrace.callFrames[0]?.lineNumber}`);
          }
        }

        // Runtime exceptions (uncaught errors)
        if (message.method === 'Runtime.exceptionThrown') {
          const { exceptionDetails } = message.params;
          log(`[ERROR] ${exceptionDetails.text}`);
          if (exceptionDetails.exception) {
            log(`  ${exceptionDetails.exception.description || JSON.stringify(exceptionDetails.exception)}`);
          }
          if (exceptionDetails.stackTrace) {
            exceptionDetails.stackTrace.callFrames.forEach(frame => {
              log(`    at ${frame.functionName || '(anonymous)'} (${frame.url}:${frame.lineNumber}:${frame.columnNumber})`);
            });
          }
        }

        // Log entries
        if (message.method === 'Log.entryAdded') {
          const { level, text, url, lineNumber } = message.params.entry;
          log(`[LOG.${level.toUpperCase()}] ${text}`);
          if (url) {
            log(`  Source: ${url}:${lineNumber}`);
          }
        }

        // Network request failures
        if (message.method === 'Network.loadingFailed') {
          const { requestId, errorText, canceled } = message.params;
          if (!canceled) {
            log(`[NETWORK FAILED] ${errorText} (Request ID: ${requestId})`);
          }
        }

        // Network responses with errors
        if (message.method === 'Network.responseReceived') {
          const { response } = message.params;
          if (response.status >= 400) {
            log(`[HTTP ${response.status}] ${response.url}`);
          }
        }

      } catch (err) {
        // Ignore parse errors
      }
    });

    ws.on('error', (err) => {
      log(`WebSocket error: ${err.message}`);
    });

    ws.on('close', () => {
      log('Connection closed. Monitoring stopped.');
      process.exit(0);
    });

  } catch (err) {
    log(`Error: ${err.message}`);
    process.exit(1);
  }
}

main();

File: livetrace/chrome-console-monitor.py
#!/usr/bin/env python3

import json
import requests
import websocket
import sys
from datetime import datetime

LOG_FILE = '/tmp/chrome-live-console.log'
DEBUG_PORT = 9222

def log(message):
    timestamp = datetime.now().isoformat()
    log_line = f"[{timestamp}] {message}\n"
    with open(LOG_FILE, 'a') as f:
        f.write(log_line)
    print(log_line.strip())

def get_websocket_debugger_url():
    try:
        response = requests.get(f'http://localhost:{DEBUG_PORT}/json')
        tabs = response.json()
        tab = next((t for t in tabs if t.get('type') == 'page'), tabs[0] if tabs else None)
        if tab and 'webSocketDebuggerUrl' in tab:
            return tab['webSocketDebuggerUrl']
        raise Exception('No debugger URL found')
    except Exception as e:
        raise Exception(f'Failed to get debugger URL: {e}')

def on_message(ws, message):
    try:
        data = json.loads(message)

        # Console API messages (console.log, console.error, etc.)
        if data.get('method') == 'Runtime.consoleAPICalled':
            params = data.get('params', {})
            msg_type = params.get('type', 'log')
            args = params.get('args', [])

            log_args = []
            for arg in args:
                if 'value' in arg:
                    log_args.append(str(arg['value']))
                elif 'description' in arg:
                    log_args.append(arg['description'])
                else:
                    log_args.append(json.dumps(arg))

            log(f"[CONSOLE.{msg_type.upper()}] {' '.join(log_args)}")

            if 'stackTrace' in params and params['stackTrace'].get('callFrames'):
                frame = params['stackTrace']['callFrames'][0]
                log(f"  Stack: {frame.get('url', 'unknown')}:{frame.get('lineNumber', '?')}")

        # Runtime exceptions (uncaught errors)
        elif data.get('method') == 'Runtime.exceptionThrown':
            details = data.get('params', {}).get('exceptionDetails', {})
            log(f"[ERROR] {details.get('text', 'Unknown error')}")

            if 'exception' in details:
                exc = details['exception']
                desc = exc.get('description', json.dumps(exc))
                log(f"  {desc}")

            if 'stackTrace' in details and details['stackTrace'].get('callFrames'):
                for frame in details['stackTrace']['callFrames']:
                    func_name = frame.get('functionName', '(anonymous)')
                    url = frame.get('url', 'unknown')
                    line = frame.get('lineNumber', '?')
                    col = frame.get('columnNumber', '?')
                    log(f"    at {func_name} ({url}:{line}:{col})")

        # Log entries
        elif data.get('method') == 'Log.entryAdded':
            entry = data.get('params', {}).get('entry', {})
            level = entry.get('level', 'info')
            text = entry.get('text', '')
            url = entry.get('url', '')
            line_number = entry.get('lineNumber', '')

            log(f"[LOG.{level.upper()}] {text}")
            if url:
                log(f"  Source: {url}:{line_number}")

        # Network request failures
        elif data.get('method') == 'Network.loadingFailed':
            params = data.get('params', {})
            if not params.get('canceled', False):
                error_text = params.get('errorText', 'Unknown error')
                request_id = params.get('requestId', 'unknown')
                log(f"[NETWORK FAILED] {error_text} (Request ID: {request_id})")

        # Network responses with errors
        elif data.get('method') == 'Network.responseReceived':
            response = data.get('params', {}).get('response', {})
            status = response.get('status', 200)
            if status >= 400:
                url = response.get('url', 'unknown')
                log(f"[HTTP {status}] {url}")

    except Exception as e:
        # Ignore parse errors
        pass

def on_error(ws, error):
    log(f"WebSocket error: {error}")

def on_close(ws, close_status_code, close_msg):
    log("Connection closed. Monitoring stopped.")

def on_open(ws):
    log("Connected to Chrome DevTools")

    # Enable Console domain
    ws.send(json.dumps({"id": 1, "method": "Console.enable"}))

    # Enable Runtime domain for console messages
    ws.send(json.dumps({"id": 2, "method": "Runtime.enable"}))

    # Enable Log domain
    ws.send(json.dumps({"id": 3, "method": "Log.enable"}))

    # Enable Network domain for network logs
    ws.send(json.dumps({"id": 4, "method": "Network.enable"}))

    log("Console monitoring enabled. Waiting for messages...\n")

def main():
    # Initialize log file
    with open(LOG_FILE, 'w') as f:
        f.write(f"Chrome Console Monitor Started - {datetime.now().isoformat()}\n")
        f.write("=" * 80 + "\n\n")

    try:
        log("Connecting to Chrome DevTools Protocol...")
        ws_url = get_websocket_debugger_url()
        log(f"WebSocket URL: {ws_url}")

        ws = websocket.WebSocketApp(
            ws_url,
            on_open=on_open,
            on_message=on_message,
            on_error=on_error,
            on_close=on_close
        )

        ws.run_forever()

    except Exception as e:
        log(f"Error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()
